component/remote/git/git.go

```diff
@@ -0,0 +1,287 @@
+
+	"github.com/go-kit/log"
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/component/remote/git/internal/vcs"
+	"github.com/grafana/agent/pkg/flow/logging/level"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name:    "remote.git",
+		Args:    Arguments{},
+		Exports: Exports{},
+
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+// Arguments configures the remote.git component.
+type Arguments struct {
+	Repository    string            `river:"repository,attr"`
+	Revision      string            `river:"revision,attr,optional"`
+	Path          string            `river:"path,attr"`
+	PullFrequency time.Duration     `river:"pull_frequency,attr,optional"`
+	GitAuthConfig vcs.GitAuthConfig `river:",squash"`
+}
+
+// Exports holds settings exported by remote.git.
+type Exports struct {
+	Content string `river:"content,attr"`
+}
+
+// DefaultArguments holds default settings for Arguments.
+var DefaultArguments = Arguments{
+	Revision:      "HEAD",
+	PullFrequency: time.Minute,
+}
+
+// SetToDefault implements river.Defaulter.
+func (args *Arguments) SetToDefault() {
+	*args = DefaultArguments
+}
+
+// Validate implements river.Validator.
+func (args *Arguments) Validate() error {
+	if args.PullFrequency < 0 {
+		return fmt.Errorf("poll_frequency cannot be negative")
+	}
+	return nil
+}
+
+// Component implements the remote.git component.
+type Component struct {
+	opts component.Options
+	log  log.Logger
+
+	mut         sync.RWMutex
+	repo        *vcs.GitRepo
+	repoOpts    vcs.GitRepoOptions
+	args        Arguments
+	lastExports Exports
+
+	argsChanged chan struct{}
+
+	healthMut sync.RWMutex
+	health    component.Health
+}
+
+var (
+	_ component.Component       = (*Component)(nil)
+	_ component.HealthComponent = (*Component)(nil)
+)
+
+// New creates a new remote.git component.
+func New(o component.Options, args Arguments) (*Component, error) {
+	c := &Component{
+		opts: o,
+		log:  o.Logger,
+
+		argsChanged: make(chan struct{}, 1),
+
+		health: component.Health{
+			Health:     component.HealthTypeUnknown,
+			Message:    "component started",
+			UpdateTime: time.Now(),
+		},
+	}
+
+	// Only acknowledge the error from Update if it's not a
+	// vcs.UpdateFailedError; vcs.UpdateFailedError means that the Git repo
+	// exists but we were just unable to update it.
+	if err := c.Update(args); err != nil {
+		if errors.As(err, &vcs.UpdateFailedError{}) {
+			level.Error(c.log).Log("msg", "failed to update repository", "err", err)
+		} else {
+			return nil, err
+		}
+	}
+	return c, nil
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	var (
+		ticker  *time.Ticker
+		tickerC <-chan time.Time
+	)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+
+		case <-c.argsChanged:
+			c.mut.Lock()
```

Comment:
For readability, you can just extract the `PullFrequency` into local variable while holding the mutex, and then the rest of the code doesn't need to hold `mut`.